OrbitQt/servicedeploymanager.cpp

```diff
@@ -0,0 +1,636 @@
+#include "OrbitSshQt/ScopedConnection.h"
+#include "OrbitSshQt/SftpChannel.h"
+#include "OrbitSshQt/SftpOperation.h"
+#include "OrbitSshQt/Task.h"
+
+static const std::string kLocalhost = "127.0.0.1";
+
+namespace OrbitQt {
+
+ServiceDeployManager::ServiceDeployManager(
+    DeploymentConfiguration deployment_configuration,
+    OrbitSsh::Credentials credentials, ServiceDeployManager::Ports remote_ports,
+    QObject* parent)
+    : QObject(parent),
+      deployment_configuration_(std::move(deployment_configuration)),
+      credentials_(std::move(credentials)),
+      remote_ports_(std::move(remote_ports)) {}
+
+void ServiceDeployManager::cancel() { loop.exit(-1); }
+
+outcome::result<bool> ServiceDeployManager::checkIfInstalled() {
+  emit statusMessage(
+      QString(
+          "Checking if OrbitService is already installed in version %1 on the "
+          "remote instance.")
+          .arg(QApplication::applicationVersion()));
+
+  auto version = QApplication::applicationVersion().toStdString();
+  if (!version.empty() && version.front() == 'v') {
+    // The old git tags have a 'v' in front which is not supported by debian
+    // packages. So we have to remove it.
+    version = version.substr(1);
+  }
+  const auto command = absl::StrFormat(
+      "/usr/bin/dpkg-query -W orbitprofiler 2>/dev/null | grep %s", version);
+
+  OrbitSshQt::Task checkIfInstalledTask{&session_.value(), command,
+                                        OrbitSshQt::Task::Tty::kNo};
+
+  QObject::connect(&checkIfInstalledTask, &OrbitSshQt::Task::finished, this,
+                   [&](int exit_code) { loop.exit(exit_code == 0 ? 0 : 1); });
+
+  std::error_code error;
+  QObject::connect(&checkIfInstalledTask, &OrbitSshQt::Task::errorOccured, this,
+                   [&](std::error_code e) {
+                     error = e;
+                     loop.exit(2);
+                   });
+
+  checkIfInstalledTask.Start();
+
+  switch (const auto result = loop.exec(); result) {
+    case 0:
+      // Already installed
+      emit statusMessage(
+          "The correct version of OrbitService is already installed.");
+      return outcome::success(true);
+    case 1:
+      emit statusMessage(
+          "The correct version of OrbitService is not yet installed.");
+      return outcome::success(false);
+    case -1:
+      return std::errc::operation_canceled;
+    case 2:
+    default:
+      return outcome::failure(error);
+  }
+}
+
+outcome::result<uint16_t> ServiceDeployManager::startAsioTunnel() {
+  if (asio_tunnel_) {
+    return outcome::success(asio_tunnel_->GetListenPort());
+  }
+
+  emit statusMessage("Setting up port forwarding...");
+  LOG("Setting up Asio tunnel on port %d", remote_ports_.asio_port);
+
+  asio_tunnel_.emplace(&session_.value(), kLocalhost, remote_ports_.asio_port);
+
+  std::error_code error;
+  OrbitSshQt::ScopedConnection errorHandler =
+      QObject::connect(&asio_tunnel_.value(), &OrbitSshQt::Tunnel::errorOccured,
```

Comment:
There are a lot of these error and quitHandler which do basically the same. Maybe it makes sense to make a template function to avoid duplicate code.

Reply:
Yes. I was somehow expecting this comment. I cleaned this up quite a bit and added our own eventloop class which works well with outcome. This shortens the code quite a bit. Thanks!